framework/src/main/java/org/checkerframework/common/wholeprograminference/WholeProgramInferenceImplementation.java

```diff
@@ -757,90 +560,90 @@ protected final void updateAnnotationSet(
-     *
-     * @param variableElement the VariableElement
-     * @return the "flatname" of the class enclosing {@code variableElement}
-     */
-    private @BinaryName String getEnclosingClassName(VariableElement variableElement) {
-        return getClassName(ElementUtils.enclosingClass(variableElement));
-    }
-
-    /**
-     * Returns the "flatname" of the class enclosing {@code localVariableNode}.
-     *
-     * @param localVariableNode the {@link LocalVariableNode}
-     * @return the "flatname" of the class enclosing {@code localVariableNode}
-     */
-    @SuppressWarnings("signature") // https://tinyurl.com/cfissue/3094
-    private @BinaryName String getEnclosingClassName(LocalVariableNode localVariableNode) {
-        return ((ClassSymbol) ElementUtils.enclosingClass(localVariableNode.getElement()))
-                .flatName()
-                .toString();
+        if (rhsATM instanceof AnnotatedNullType && ignoreNullAssignments) {
+            return;
+        }
+        AnnotatedTypeMirror atmFromStorage =
+                storage.atmFromAnnotationLocation(rhsATM.getUnderlyingType(), typeToUpdate);
+        updateATMWithLUB(rhsATM, atmFromStorage);
+        if (lhsATM instanceof AnnotatedTypeVariable) {
+            Set<AnnotationMirror> upperAnnos =
+                    ((AnnotatedTypeVariable) lhsATM).getUpperBound().getEffectiveAnnotations();
+            // If the inferred type is a subtype of the upper bounds of the
+            // current type on the source code, halt.
+            if (upperAnnos.size() == rhsATM.getAnnotations().size()
+                    && atypeFactory
+                            .getQualifierHierarchy()
+                            .isSubtype(rhsATM.getAnnotations(), upperAnnos)) {
+                return;
+            }
+        }
+        storage.updateStorageLocationFromAtm(
+                rhsATM, lhsATM, typeToUpdate, defLoc, ignoreIfAnnotated);
+        storage.setFileModified(file);
     }
 
     /**
-     * Returns the top-level class that contains {@code element}.
+     * Updates sourceCodeATM to contain the LUB between sourceCodeATM and ajavaATM, ignoring missing
+     * AnnotationMirrors from ajavaATM -- it considers the LUB between an AnnotationMirror am and a
+     * missing AnnotationMirror to be am. The results are stored in sourceCodeATM.
      *
-     * @param element the element wose enclosing class to find
-     * @return an element for a class containing {@code element} that isn't contained in another
-     *     class
+     * @param sourceCodeATM the annotated type on the source code
+     * @param ajavaATM the annotated type on the ajava file
      */
-    private TypeElement toplevelEnclosingClass(Element element) {
-        if (ElementUtils.enclosingClass(element) == null) {
-            return (TypeElement) element;
+    private void updateATMWithLUB(AnnotatedTypeMirror sourceCodeATM, AnnotatedTypeMirror ajavaATM) {
```

Comment:
The capitalization in this method name is hard to read.  Style guides I know of recommend downcasing letters after the first in an abbreviation, when used in an identifier.  That would be "updateAtmWithLub".  An alternate approach would be to change "ATM" here to "Type", which is less implementation-specific.